src/TestFramework/TestFramework/Assertions/CollectionAssert.cs

```diff
@@ -1605,43 +1605,57 @@ internal static bool IsSubsetOfHelper(ICollection subset, ICollection superset)
         ref string reason)
     {
         Assert.CheckParameterNotNull(comparer, "Assert.AreCollectionsEqual", "comparer", string.Empty);
-        if (!ReferenceEquals(expected, actual))
+        if (ReferenceEquals(expected, actual))
         {
-            if ((expected == null) || (actual == null))
-            {
-                return false;
-            }
+            reason = string.Format(CultureInfo.CurrentCulture, FrameworkMessages.BothCollectionsSameReference, string.Empty);
+            return true;
+        }
 
-            if (expected.Count != actual.Count)
-            {
-                reason = FrameworkMessages.NumberOfElementsDiff;
-                return false;
-            }
+        return CompareCollections(expected, actual, comparer, ref reason);
+    }
+
+    private static bool CompareCollections(ICollection? expected, ICollection? actual, IComparer comparer, ref string reason)
+    {
+        if ((expected == null) || (actual == null))
+        {
+            return false;
+        }
 
-            IEnumerator expectedEnum = expected.GetEnumerator();
-            IEnumerator actualEnum = actual.GetEnumerator();
-            int i = 0;
-            while (expectedEnum.MoveNext() && actualEnum.MoveNext())
+        if (expected.Count != actual.Count)
+        {
+            reason = FrameworkMessages.NumberOfElementsDiff;
+            return false;
+        }
+
+        IEnumerator expectedEnum = expected.GetEnumerator();
+        IEnumerator actualEnum = actual.GetEnumerator();
+
+        int position = 0;
+        while (expectedEnum.MoveNext() && actualEnum.MoveNext())
+        {
+            object? curExpected = expectedEnum.Current;
+            object? curActual = actualEnum.Current;
+
+            if (curExpected is ICollection curExpectedCollection && curActual is ICollection curActualCollection)
```

Comment:
We may want to allow IEnumerable because I think your fix wouldn't work for a non-ICollection inner collection. You could try with `List<ReadOnlyCollection<int>>`